framework/src/utils/RandomData.C

```diff
@@ -91,32 +102,55 @@ RandomData::updateGenerators()
     _generator.seed(MASTER, parallel_seed);
   }
 
-  if (_is_nodal)
-    updateGeneratorHelper(_rd_mesh.getMesh().active_nodes_begin(),
-                          _rd_mesh.getMesh().active_nodes_end());
-  else
-    updateGeneratorHelper(_rd_mesh.getMesh().active_elements_begin(),
-                          _rd_mesh.getMesh().active_elements_end());
-}
-
-template <typename T>
-void
-RandomData::updateGeneratorHelper(T it, T end_it)
-{
-  processor_id_type processor_id = _rd_problem.processor_id();
+  auto processor_id = _rd_problem.processor_id();
 
-  for (; it != end_it; ++it)
+  if (_is_nodal)
   {
-    dof_id_type id = (*it)->id();
-    uint32_t rand_int = _generator.randl(MASTER);
+    const auto & node_to_elem = _rd_mesh.nodeToElemMap();
+    auto & mesh = _rd_mesh.getMesh();
 
-    if (processor_id == (*it)->processor_id())
+    for (const auto node_ptr :
+         as_range(_rd_mesh.getMesh().active_nodes_begin(), _rd_mesh.getMesh().active_nodes_end()))
+    {
+      auto id = node_ptr->id();
+      auto rand_int = _generator.randl(MASTER);
+
+      // Only save states for nodes attached to active elements
+      auto elem_id_it = node_to_elem.find(id);
+
+      if (elem_id_it != node_to_elem.end())
+      {
+        for (auto elem_id : elem_id_it->second)
+        {
+          const auto * elem_ptr = mesh.elem_ptr(elem_id);
+
+          if (elem_ptr && processor_id == elem_ptr->processor_id())
+          {
+            _seeds[id] = rand_int;
+
+            // Update the individual dof object generators
+            _generator.seed(id, rand_int);
+          }
+        }
+      }
+    }
+  }
+  else
+  {
+    for (const auto & elem_ptr : as_range(_rd_mesh.getMesh().active_elements_begin(),
+                                          _rd_mesh.getMesh().active_elements_end()))
     {
-      // Only save states for local dofs
-      _seeds[id] = rand_int;
+      auto id = elem_ptr->id();
+      auto rand_int = _generator.randl(MASTER);
```

Comment:
You are effectively already _forwarding_ the master RNG by reading the stream of numbers (and throwing lots of them away). The scheme we talked about yesterday would be no worse in that respect, but it would be parallel agnostic with distributed mesh as well.

Reply:
Yes, I agree. I'm going on vacation and when I get back will be hitting another project pretty hard. This is probably going to go in with minor corrections or just be closed. Just too busy...